File: yggdrasil/examples/formatted_io1/src/formatted_io1_modelA.c
#include <stdio.h>
// Include methods for input/output channels
#include "YggInterface.h"

#define MYBUFSIZ 1000

int main(int argc, char *argv[]) {
  // Initialize input/output channels
  yggInput_t in_channel = yggInput("inputA");
  yggOutput_t out_channel = yggOutput("outputA");

  // Declare resulting variables and create buffer for received message
  int flag = 1;
  size_t msg_siz = 0;
  char *msg = NULL;

  // Loop until there is no longer input or the queues are closed
  while (flag >= 0) {

    // Receive input from input channel
    // If there is an error, the flag will be negative
    // Otherwise, it is the number of variables filled
    flag = yggRecvRealloc(in_channel, &msg, &msg_siz);
    if (flag < 0) {
      printf("Model A: No more input.\n");
      break;
    }

    // Print received message
    printf("Model A: %s\n", msg);

    // Send output to output channel
    // If there is an error, the flag will be negative
    flag = yggSend(out_channel, msg, msg_siz);
    if (flag < 0) {
      printf("Model A: Error sending output.\n");
      break;
    }

  }
  
  return 0;
}

File: yggdrasil/examples/formatted_io1/src/formatted_io1_modelA.f90
program main
  ! Include methods for input/output channels
  use fygg

  ! Declare resulting variables and create buffer for received message
  integer, parameter :: mybufsiz = 1000
  integer :: flag = 1
  type(yggcomm) :: in_channel, out_channel
  character(len=100), target :: msg = "hello"
  ! character(len=:), allocatable, target :: msg
  ! allocate(character(len=100) :: msg)
  integer(kind=c_size_t), target :: msg_siz = 100

  ! initialize input/output channels
  in_channel = ygg_input("inputA")
  out_channel = ygg_output("outputA")

  ! Loop until there is no longer input or the queues are closed
  do while (flag.ge.0)

     ! Receive input from input channel
     ! If there is an error, the flag will be negative
     ! Otherwise, it is the number of variables filled
     flag = ygg_recv_var(in_channel, &
          [yggarg(msg), yggarg(msg_siz)])
     if (flag.lt.0) then
        print *, "Model A: No more input."
        exit
     end if

     ! Print received message
     print *, "Model A: ", msg

     ! Send output to output channel
     ! If there is an error, the flag will be negative
     flag = ygg_send_var(out_channel, &
          [yggarg(msg), yggarg(msg_siz)])
     if (flag.lt.0) then
        print *, "Model A: Error sending output."
        exit
     end if

  end do

  call exit(0)

end program main
